File: TPII-2025-2/PadraoGoF/Estrutural/Flyweight/Flyweigth01.js
//Flyweight factory
class FabricaFlyweight{
    constructor(){
        this.flyweights = {};
    }

    obterFlyweigth(id){
        if(!this.flyweights[id]){
            this.flyweights[id] = new ProdutoFlyweigth(id);
        }
        return this.flyweights[id];
    }
}
//Flyweight
class ProdutoFlyweigth{
    constructor(id){
        this.id = id;
    }

    exibir(nome,preco){
        console.log(`ID: ${this.id} | R$${preco.toFixed(2)} | PROD: ${nome}`);
    }
}

//Uso padrão
class Cliente{
    constructor(){
        this.fabricaFlyweight = new FabricaFlyweight();
        this.pedido = [];
    }

    addProduto(id, nome, preco){
        const flyweights = this.fabricaFlyweight.obterFlyweigth(id);
        this.pedido.push({flyweights, nome, preco});
    }

    exibirPedido(){
        console.log("Itens do pedido:")
        this.pedido.forEach(item => {
            item.flyweights.exibir(item.nome, item.preco);
        })
    }
}

//Aplicação
let valor = 99.99;
const cliente = new Cliente();

cliente.addProduto("001", "Camiseta", 39.90);
cliente.addProduto("002", "Touca", 19.90);
cliente.addProduto("003", "Meia", 59.90);

cliente.addProduto("002", "Chinelo", valor);
cliente.addProduto("003", "Boné", 49.90);

cliente.exibirPedido();

File: TPII-2025-2/PadraoGoF/Estrutural/Flyweight/Flyweigth01.py
#Flyweight Factory
class FabricaFlyweigth:
    def __init__(self):
        self.flyweights = {}
    
    def obterFlyweigth(self, id):
        if id not in self.flyweights:
            self.flyweights[id] = ProdutoFlyweigth(id)
        return self.flyweights[id]

#Flyweigth
class ProdutoFlyweigth:
    def __init__(self, id):
        self.id = id
    
    def exibir(self, nome, preco):
        print(f"ID: {self.id} | R${preco: .2f} | PROD: {nome}")

#Uso padrao
class Cliente:
    def __init__ (self):
        self.fabricaFlyweigth = FabricaFlyweigth()
        self.pedido = []

    def addProduto(self, id, nome, preco):
        FLYWEIGTHS = self.fabricaFlyweigth.obterFlyweigth(id)
        self.pedido.append({"flyweights": FLYWEIGTHS, "nome": nome, "preco": preco})

    def exibirPedido(self):
        print("Itens do pedido:")
        for item in self.pedido:
             item["flyweights"].exibir(item["nome"], item["preco"])

#Aplicação
valor = 99.99
CLIENTE = Cliente()

CLIENTE.addProduto("001", "Camiseta", 39.90)
CLIENTE.addProduto("002", "Touca", 19.90)
CLIENTE.addProduto("003", "Meia", 59.90)

CLIENTE.addProduto("002", "Chinelo", valor)
CLIENTE.addProduto("003", "Boeeeené", 49.90)

CLIENTE.exibirPedido()